components/brave_wallet/browser/ethereum_permission_utils.cc

```diff
@@ -0,0 +1,158 @@
+
+#include "base/no_destructor.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "base/strings/stringprintf.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "url/gurl.h"
+
+namespace {
+
+constexpr char kAddrPattern[] = "addr%3D(0x[[:xdigit:]]{40})";
+
+// Given an origin and an account address, append the account address to the
+// end of the host piece of the origin, then return it as the new origin.
+bool AddAccountToHost(const GURL& old_origin,
+                      const std::string& account,
+                      GURL* new_origin) {
+  if (!old_origin.is_valid() || account.empty() || !new_origin)
+    return false;
+
+  GURL::Replacements replacements;
+  std::string new_host = base::StrCat({old_origin.host_piece(), account});
+  replacements.SetHostStr(new_host);
+
+  *new_origin = old_origin.ReplaceComponents(replacements);
+  return new_origin->is_valid();
+}
+
+// Given the overwritten origin, such as https://test.com{addr=123&addr=456},
+// extract all addresses and save into address_queue.
+void ExtractAddresses(const GURL& origin,
+                      std::queue<std::string>* address_queue) {
+  static const base::NoDestructor<re2::RE2> kAddrRegex(kAddrPattern);
+  DCHECK(origin.is_valid() && address_queue);
+  DCHECK_EQ(origin, origin.GetOrigin());
+
+  re2::StringPiece input(origin.spec());
+  std::string match;
+  while (re2::RE2::FindAndConsume(&input, *kAddrRegex, &match)) {
+    address_queue->push(match);
+  }
+}
+
+// Parse requesting origin in either sub-request format (one address) or
+// non-sub-request format (all addresses).
+bool ParseRequestingOriginInternal(const GURL& origin,
+                                   bool sub_req_format,
+                                   std::string* requesting_origin,
+                                   std::string* account,
+                                   std::queue<std::string>* address_queue) {
+  if (!origin.is_valid() || origin != origin.GetOrigin())
+    return false;
+
+  std::string scheme_host_group;
+  std::string address_group;
+  std::string port_group;
+
+  // Validate input format.
+  std::string pattern = sub_req_format
+                            ? "(.*)(0x[[:xdigit:]]{40})(:[0-9]+)*/"
+                            : "(.*)%7Baddr%3D0x[[:xdigit:]]{40}(%"
+                              "26addr%3D0x[[:xdigit:]]{40})*%7D(:[0-9]+)*/";
+  RE2 full_pattern(pattern);
+  if (!re2::RE2::FullMatch(origin.spec(), full_pattern, &scheme_host_group,
+                           &address_group, &port_group)) {
+    return false;
+  }
+
+  if (requesting_origin) {
+    *requesting_origin = base::StrCat({scheme_host_group, port_group});
+  }
+
+  if (sub_req_format && account) {
+    *account = address_group;
+  }
+
+  if (!sub_req_format && address_queue) {
+    ExtractAddresses(origin, address_queue);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+bool GetConcatOriginFromWalletAddresses(
```

Comment:
`absl::optional` is preferred over `bool` with an out param

Reply:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/abseil-cpp/absl/types/optional.h;l=23?q=optional.h&ss=chromium

Reply:
https://github.com/brave/brave-browser/issues/17270